src/man/man1/mbformat.1
.TH mbformat 1 "7 September 2026" "MB-System 5.0" "MB-System 5.0"
.SH NAME
\fBmbformat\fP \-  List information about swath sonar data formats
supported by the \fBMBIO\fP library.

.SH VERSION
Version 5.0

.SH SYNOPSIS
\fBmbformat\fP
.br
[--format=format_id {-Fformat_id}
.br
--input=file {-Ifile}
.br
--list-ids {-L}
.br
--list-root {-K}
.br
--verbose {-V}
.br
--html {-W}
.br
--help {-H}]

.SH DESCRIPTION
\fBmbformat\fP is a utility which identifies the swath sonar data formats
associated with \fBmbio\fP format id's.  If no format id or input filename
is specified, \fBmbformat\fP lists all of the currently supported formats 
along with short descriptions. If a format id is specified using the 
\fB\-F\fP option, then \fBmbformat\fP prints a short description of that 
format. If a filename is specified using the \fB\-I\fP option, then 
\fBmbformat\fP attempts to identify a valid format id for that file using 
swath data filename prefix and suffix conventions embedded in \fBMB-System\fP. 
The program also determines the filename "root", which is the filename minus 
whatever format-specific prefix or suffix has been recognized. In no prefix
or suffix is recognized, then the root is simply the entire filename.

Normally, \fBmbformat\fP
prints out a brief description of each identified format. If the \fB\-L\fP
option is used, \fBmbformat\fP simply prints out the valid format id number,
or 0 if a file or format has been specified but no valid id can be identified.
If the \fB\-K\fP option is used, \fBmbformat\fP will print out the filename
"root" followed by the format id number.

.SH MB-SYSTEM AUTHORSHIP
David W. Caress
.br
  Monterey Bay Aquarium Research Institute
.br
Dale N. Chayes
.br
  Center for Coastal and Ocean Mapping
.br
  University of New Hampshire
.br
Christian do Santos Ferreira
.br
  MARUM - Center for Marine Environmental Sciences
.br
  University of Bremen

.SH OPTIONS
.TP
.B \-\-format=format_id \-Fformat_id
.br
Sets the format id to be listed.  If no format is given, all of the
currently supported formats will be listed. A list of the swath sonar data formats
currently supported by \fBMBIO\fP and their identifier values
is given in the \fBMBIO\fP manual page.
.TP
.B \-\-help \-H
This "help" flag cause the program to print out a description
of its operation and then exit immediately.
.TP
.B \-\-input=file \-Ifile
.br
Sets the filename to be tested against known filename prefix and
suffix conventions in order to identify the format id.
.TP
.B \-\-list\-root \-K
Causes the fileroot and the format id to be output. The fileroot
is the filename with any format specific suffix removed.
.TP
.B \-\-list\-ids \-L
Causes the format descriptions to be omitted
so that only the \fBMBIO\fP format id numbers are listed.
.TP
.B \-\-verbose \-V
Normally, \fBmbformat\fP only prints out format descriptions.  If the
\fB\-V\fP flag is given, then \fBmbformat\fP works in a "verbose" mode and
also outputs the program version being used.
.TP
.B \-\-html \-W
Output the list of formats as HTML.

.SH EXAMPLES
Suppose one wishes to identify the swath sonar data format associated
with a format id of 11.  The following will suffice:
.br
 	mbformat \-F11
.br
and the result will be this output:
  MBIO data format id: 11
  Format name:          MBF_SBSIOMRG
  Informal Description: SIO merge Sea Beam
  Attributes:           Sea Beam, bathymetry, 16 beams, binary, uncentered,
                        SIO.

Suppose one wishes to list all of the currently supported data formats.
The following will suffice:
.br
 	mbformat

Suppose one wants to find out what format is associated with a raw
SeaBeam 2112 file named sb199811231045.rec:
.br
 	mbformat \-I sb199811231045.rec
.br
This results in the output:
  MBIO data format id: 41
  Format name:          MBF_SB2100RW
  Informal Description: SeaBeam 2100 series vender format
  Attributes:           SeaBeam 2100, bathymetry, amplitude 
                        and sidescan, 151 beams and 2000 pixels, ascii 
                        with binary sidescan, SeaBeam Instruments.

Using the \-K option:
.br
 	mbformat \-K \-I sb199811231045.rec
.br
will produce the following output:
.br
  sb199811231045 41
.br
where the format id has been identified as 41 because of the ".rec"
file suffix and the fileroot has been identified as "sb199811231045"
by removing the suffix. Use of the \-L option:
.br
 	mbformat \-L \-I sb199811231045.rec
.br
will produce output with just format id:
.br
  41

.SH SEE ALSO
\fBmbsystem\fP(1), \fBmbio\fP(1)

.SH BUGS
No known bugs.


NEWS
-------------------------------------------------------------------------------
MB-SYSTEM NEWS FILE:
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------

In the GNU package standard, the NEWS file lists the history of user-visible
changes to a package. For the the MB-System open source software package for the
processing and display of swath sonar data, this information is included in the
ChangeLog file. Both the NEWS file and the ChangeLog file are located at the top
of the MB-System source code distribution directory structure.

-------------------------------------------------------------------------------


src/utilities/mbformat.cc
/*--------------------------------------------------------------------
 *    The MB-system:	mbformat.c	1/22/93
 *
 *    Copyright (c) 1993-2025 by
 *    David W. Caress (caress@mbari.org)
 *      Monterey Bay Aquarium Research Institute
 *      Moss Landing, California, USA
 *    Dale N. Chayes 
 *      Center for Coastal and Ocean Mapping
 *      University of New Hampshire
 *      Durham, New Hampshire, USA
 *    Christian dos Santos Ferreira
 *      MARUM
 *      University of Bremen
 *      Bremen Germany
 *     
 *    MB-System was created by Caress and Chayes in 1992 at the
 *      Lamont-Doherty Earth Observatory
 *      Columbia University
 *      Palisades, NY 10964
 *
 *    See README.md file for copying and redistribution conditions.
 *--------------------------------------------------------------------*/
/*
 * MBFORMAT provides a description of the swath data format
 * associated with a particular MBIO format identifier.  If
 * no format is specified, MBFORMAT will list descriptions of all
 * the currently supported formats.
 *
 * Author:	D. W. Caress
 * Date:	January 22, 1993
 */

#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#ifdef _WIN32
#include "unistd_w.h"
#else
#include <unistd.h>
#endif
#include "mb_define.h"
#include "mb_format.h"
#include "mb_status.h"

enum MbformatList {
  MBFORMAT_LIST_LONG = 0,
  MBFORMAT_LIST_SIMPLE = 1,
  MBFORMAT_LIST_ROOT = 2
};

constexpr char program_name[] = "MBFORMAT";
constexpr char help_message[] =
    "MBFORMAT is an utility which identifies the swath data formats\n"
    "associated with MBIO format id's.  If no format id is specified,\n"
    "MBFORMAT lists all of the currently supported formats.";
constexpr char usage_message_old[] = "mbformat [-Fformat -Ifile -L -K -V -W -H]";
constexpr char usage_message[] =
    "mbformat\n"
    "\t--format=format_id {-Fformat_id}\n"
    "\t--help {-H}\n"
    "\t--html {-W}\n"
    "\t--input=file {-Ifile}\n"
    "\t--list-ids {-L}\n"
    "\t--list-root {-K}\n"
    "\t--verbose {-V}\n";

/*--------------------------------------------------------------------*/
int main(int argc, char **argv) {
	char file[MB_PATH_MAXLINE];
	int verbose = 0;
	bool file_specified = false;
	int format = 0;
	bool format_specified = false;
	bool html = false;
	enum MbformatList list_mode = MBFORMAT_LIST_LONG;

	/* process argument list */
	{
		static struct option options[] = {{"format", required_argument, nullptr, 0},
		                                  {"help", no_argument, nullptr, 0},
		                                  {"html", no_argument, nullptr, 0},
		                                  {"input", required_argument, nullptr, 0},
		                                  {"list-ids", no_argument, nullptr, 0},
		                                  {"list-root", no_argument, nullptr, 0},
		                                  {"verbose", no_argument, nullptr, 0},
		                                  {nullptr, 0, nullptr, 0}};

		bool errflg = false;
		bool help = false;
		int c;
		int option_index;
		while ((c = getopt_long(argc, argv, "F:f:HhI:i:LlKkVvWw", options, &option_index)) != -1)
			switch (c) {
			case 0:
				if (strcmp("format", options[option_index].name) == 0) {
					sscanf(optarg, "%d", &format);
					format_specified = true;
				}
				else if (strcmp("list-ids", options[option_index].name) == 0) {
					list_mode = MBFORMAT_LIST_SIMPLE;
				}
				else if (strcmp("list-root", options[option_index].name) == 0) {
					list_mode = MBFORMAT_LIST_ROOT;
				}
				else if (strcmp("help", options[option_index].name) == 0) {
					help = true;
				}
				else if (strcmp("input", options[option_index].name) == 0) {
					sscanf(optarg, "%1023s", file);
					file_specified = true;
				}
				else if (strcmp("verbose", options[option_index].name) == 0) {
					verbose++;
				}
				else if (strcmp("html", options[option_index].name) == 0) {
					html = true;
				}
				break;
			case 'F':
			case 'f':
				sscanf(optarg, "%d", &format);
				format_specified = true;
				break;
			case 'L':
			case 'l':
				list_mode = MBFORMAT_LIST_SIMPLE;
				break;
			case 'K':
			case 'k':
				list_mode = MBFORMAT_LIST_ROOT;
				break;
			case 'H':
			case 'h':
				help = true;
				break;
			case 'I':
			case 'i':
				sscanf(optarg, "%1023s", file);
				file_specified = true;
				break;
			case 'V':
			case 'v':
				verbose++;
				break;
			case 'W':
			case 'w':
				html = true;
				break;
			case '?':
				errflg = true;
			}

		if (errflg) {
			fprintf(stderr, "usage: %s\n", usage_message);
			exit(MB_ERROR_BAD_USAGE);
		}

		if (verbose == 1 || help) {
			fprintf(stderr, "\nProgram %s\n", program_name);
			fprintf(stderr, "MB-system Version %s\n", MB_VERSION);
		}

		if (verbose >= 2) {
			fprintf(stderr, "\ndbg2  Program <%s>\n", program_name);
			fprintf(stderr, "dbg2  MB-system Version %s\n", MB_VERSION);
			fprintf(stderr, "dbg2  Control Parameters:\n");
			fprintf(stderr, "dbg2       verbose: %d\n", verbose);
			fprintf(stderr, "dbg2       help:    %d\n", help);
			if (format_specified)
				fprintf(stderr, "dbg2       format:  %d\n", format);
			if (file_specified)
				fprintf(stderr, "dbg2       file:    %s\n", file);
		}

		if (help) {
			fprintf(stderr, "\n%s\n", help_message);
			fprintf(stderr, "\nusage: %s\n", usage_message);
			exit(MB_ERROR_NO_ERROR);
		}
	} /* process argument list */

	int status = MB_SUCCESS;
	int error = MB_ERROR_NO_ERROR;

	/* print out the info */
	int format_save = format;
	char root[MB_PATH_MAXLINE];
	if (file_specified) {
		status = mb_get_format(verbose, file, root, &format, &error);
	}
	else if (format_specified) {
		status = mb_format(verbose, &format, &error);
	}

	if (file_specified && format == 0) {
		if (list_mode == MBFORMAT_LIST_SIMPLE)
			printf("%d\n", format);
		else if (list_mode == MBFORMAT_LIST_ROOT)
			printf("%s %d\n", root, format);
		else
			printf("Program %s unable to infer format from filename %s\n", program_name, file);
	}
	else if (format_specified && format == 0) {
		if (list_mode == MBFORMAT_LIST_SIMPLE)
			printf("%d\n", format);
		else if (list_mode == MBFORMAT_LIST_ROOT)
			printf("%s %d\n", root, format);
		else
			printf("Specified format %d invalid for MB-System\n", format_save);
	}
	else if (format != 0) {
		if (list_mode == MBFORMAT_LIST_SIMPLE) {
			printf("%d\n", format);
		}
		else if (list_mode == MBFORMAT_LIST_ROOT) {
			printf("%s %d\n", root, format);
		}
		else {
			char format_description[MB_DESCRIPTION_LENGTH];
			status = mb_format_description(verbose, &format, format_description, &error);
			if (status == MB_SUCCESS) {
				printf("\nMBIO data format id: %d\n", format);
				printf("%s", format_description);
			}
			else if (file_specified) {
				printf("Program %s unable to infer format from filename %s\n", program_name, file);
			}
			else if (format_specified) {
				printf("Specified format %d invalid for MB-System\n", format_save);
			}
		}
	}
	else if (html) {
		printf("<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 3.2//EN\">\n");
		printf("<HTML>\n<HEAD>\n   <TITLE>MB-System Supported Data Formats</TITLE>\n");
		printf("</HEAD>\n<BODY TEXT=\"#000000\" BGCOLOR=\"#FFFFFF\" LINK=\"#336699\" VLINK=\"#997040\" ALINK=\"#CC9900\">\n\n");
		printf("<CENTER><P><B><FONT SIZE=+2>MB-System Supported Swath Data Formats</FONT></B></P></CENTER>\n\n");
		printf("<P>Each swath mapping sonar system outputs a data stream which includes\n");
		printf("some values or parameters unique to that system. In general, a number of\n");
		printf("different data formats have come into use for data from each of the sonar\n");
		printf("systems; many of these formats include only a subset of the original data\n");
		printf("stream. Internally, MBIO recognizes which sonar system each data format\n");
		printf("is associated with and uses a data structure including the complete data\n");
		printf("stream for that sonar. At present, formats associated with the following\n");
		printf("sonars are supported: </P>\n\n");
		printf("<UL>\n<LI>Sea Beam &quot;classic&quot; multibeam sonar </LI>\n\n");
		printf("<LI>Hydrosweep DS multibeam sonar </LI>\n\n");
		printf("<LI>Hydrosweep DS2 multibeam sonar </LI>\n\n");
		printf("<LI>Hydrosweep MD multibeam sonar </LI>\n\n");
		printf("<LI>Sea Beam 2000 multibeam sonar </LI>\n\n");
		printf("<LI>Sea Beam 2112 and 2136 multibeam sonars </LI>\n\n");
		printf("<LI>Sea Beam 2120 multibeam sonars </LI>\n\n");
		printf("<LI>Simrad EM12, EM121, EM950, and EM1000 multibeam sonars </LI>\n\n");
		printf("<LI>Simrad EM120, EM300, and EM3000 multibeam sonars</LI>\n\n");
		printf("<LI>Simrad EM122, EM302, EM710, and EM3002 multibeam sonars</LI>\n\n");
		printf("<LI>Simrad Mesotech SM2000 multibeam sonar</LI>\n\n");
		printf("<LI>Hawaii MR-1 shallow tow interferometric sonar </LI>\n\n");
		printf("<LI>ELAC Bottomchart and Bottomchart MkII shallow water multibeam sonars</LI>\n\n");
		printf("<LI>Reson Seabat multibeam sonars (e.g. 9001, 8081, 7125)</LI>\n\n");
		printf("<LI>WHOI DSL AMS-120 deep tow interferometric sonar </LI>\n\n");
		printf("<LI>Sea Scan sidescan sonar</LI>\n\n");
		printf("<LI>Furuno HS-1 multibeam sonar</LI>\n\n");
		printf("<LI>Edgetech sidescan and subbottom profiler sonars</LI>\n\n");
		printf("<LI>Imagenex DeltaT multibeam sonars</LI>\n\n");
		printf("<LI>Odom ES3 multibeam sonar</LI>\n\n");
		printf("</UL>\n\n");
		printf("<P>The following swath mapping sonar data formats are currently supported by MB-System:</P>\n\n");

		for (int i = 0; i <= 1000; i++) {
			format = i;
			char format_description[MB_DESCRIPTION_LENGTH];
			if ((status = mb_format_description(verbose, &format, format_description, &error)) == MB_SUCCESS && format == i) {
				const char *format_informal_ptr =
				    (char *)strstr(format_description, "Informal Description:");
				const char *format_attributes_ptr =
				    (char *)strstr(format_description, "Attributes:");
				if (format_informal_ptr == nullptr || format_attributes_ptr == nullptr ||
				    format_attributes_ptr < format_informal_ptr) {
					continue;
				}
				char format_name[MB_DESCRIPTION_LENGTH];
        		size_t format_name_len = MIN(MB_DESCRIPTION_LENGTH, strlen(format_description) - strlen(format_informal_ptr));
				strncpy(format_name, format_description, MIN(format_name_len, sizeof(format_description)));
				format_name[format_name_len - 1] = '\0';
				char format_informal[MB_DESCRIPTION_LENGTH];
        		size_t format_informal_len = MIN(MB_DESCRIPTION_LENGTH, strlen(format_informal_ptr) - strlen(format_attributes_ptr));
				strncpy(format_informal, format_informal_ptr, MIN(format_informal_len, sizeof(format_description)));
				format_informal[format_informal_len - 1] = '\0';
				char format_attributes[MB_DESCRIPTION_LENGTH];
				strcpy(format_attributes, format_attributes_ptr);
				format_attributes[strlen(format_attributes_ptr) - 1] = '\0';
				printf("\n<UL>\n<LI>MBIO Data Format ID:  %d </LI>\n", format);
				printf("\n<UL>\n<LI>%s</LI>\n", format_name);
				printf("\n<LI>%s</LI>\n", format_informal);
				printf("\n<LI>%s</LI>\n", format_attributes);
				printf("</UL>\n</UL>\n");
			}
		}

		printf("\n<CENTER><P><BR>\n");
		printf("\n<P>\n<HR WIDTH=\"67%%\"></P>\n\n");
    printf("<center>\n");
    printf("\t<a href=\"https://www.mbari.org/products/research-software/mb-system/\"><< MB-System website</a> ");
    printf("| <a href=\"https://www.mbari.org/technology/mb-system/documentation/\"> Online MB-System Documenation>></a> ");
    printf("| <a href=\"index.html\">MB-System Information in Local Installation</a></p>\n");
    printf("</center>\n");
		printf("\n</BODY>\n</HTML>\n");

		status = MB_SUCCESS;
		error = MB_ERROR_NO_ERROR;
	}
	else if (list_mode) {
		for (int i = 0; i <= 1000; i++) {
			format = i;
			if ((status = mb_format(verbose, &format, &error)) == MB_SUCCESS && format == i) {
				printf("%d\n", format);
			}
		}
		status = MB_SUCCESS;
		error = MB_ERROR_NO_ERROR;
	}
	else {
		printf("\nSupported MBIO Formats:\n");
		for (int i = 0; i <= 1000; i++) {
			format = i;
			char format_description[MB_DESCRIPTION_LENGTH];
			if ((status = mb_format_description(verbose, &format, format_description, &error)) == MB_SUCCESS && format == i) {
				printf("\nMBIO Data Format ID:  %d\n", format);
				printf("%s", format_description);
			}
		}
		status = MB_SUCCESS;
		error = MB_ERROR_NO_ERROR;
	}

	if (verbose >= 2) {
		fprintf(stderr, "\ndbg2  Program <%s> completed\n", program_name);
		fprintf(stderr, "dbg2  Ending status:\n");
		fprintf(stderr, "dbg2       status:  %d\n", status);
	}

	exit(error);
}
/*--------------------------------------------------------------------*/
